src/clj/puppetlabs/services/jruby/jruby_puppet_agents.clj

```diff
@@ -65,62 +65,101 @@
+  [pool-context :- jruby-schemas/PoolContext
+   old-pool :- jruby-schemas/PoolState
+   new-pool-state :- jruby-schemas/PoolState
+   refill :- schema/Bool]
   (let [{:keys [config profiler pool-state]} pool-context
-        new-pool-state (jruby-internal/create-pool-from-config config)
         new-pool (:pool new-pool-state)
-        old-pool @pool-state
-        count    (:size old-pool)]
+        old-pool-size (:size old-pool)]
     (log/info "Replacing old JRuby pool with new instance.")
     (reset! pool-state new-pool-state)
     (log/info "Swapped JRuby pools, beginning cleanup of old pool.")
-    (doseq [i (range count)]
+    (doseq [i (range old-pool-size)]
       (try
-        (let [id        (inc i)
-              instance  (jruby-internal/borrow-from-pool!*
-                          jruby-internal/borrow-without-timeout-fn
-                          (:pool old-pool))]
+        (let [id (inc i)
+              instance (jruby-internal/borrow-from-pool!*
+                        jruby-internal/borrow-without-timeout-fn
+                        (:pool old-pool))]
           (try
-            (flush-instance! pool-context instance new-pool id config profiler)
+            (jruby-internal/cleanup-pool-instance! instance)
+            (when refill
+              (jruby-internal/create-pool-instance! new-pool id config
+                                                    (partial send-flush-instance! pool-context)
+                                                    profiler)
+              (log/infof "Finished creating JRubyPuppet instance %d of %d"
+                         id old-pool-size))
             (finally
-              (.releaseItem (:pool old-pool) instance false)))
-          (log/infof "Finished creating JRubyPuppet instance %d of %d"
-                     id count))
+              (.releaseItem (:pool old-pool) instance false))))
         (catch Exception e
           (.clear new-pool)
           (.insertPill new-pool (PoisonPill. e))
           (throw (IllegalStateException.
-                   "There was a problem adding a JRubyPuppet instance to the pool."
-                   e)))))
+                  "There was a problem adding a JRubyPuppet instance to the pool."
+                  e)))))
     ;; Add a "RetryPoisonPill" to the pool in case something else is in the
     ;; process of borrowing from the old pool.
     (.insertPill (:pool old-pool) (RetryPoisonPill. (:pool old-pool)))))
 
+(schema/defn ^:always-validate
+  flush-pool-for-shutdown!
+  "Flush of the current JRuby pool when shutting down during a stop."
+  ;; Since this function is only called by the pool-agent, we know that if we
+  ;; receive multiple flush requests before the first one finishes, they will
+  ;; be queued up and we don't need to worry about race conditions between the
+  ;; steps we perform here in the body.
+  [pool-context :- jruby-schemas/PoolContext
+   flush-complete? :- IDeref]
```

Comment:
It's probably better to name this something like `on-complete` to indicate that it's expected to be a promise and not a boolean.  Also, could use a mention in the docstring.